File: Python/Problem_062.py
def euler_62() -> int:
    digits_len = 1
    while True:
        seen = {}  # sig -> [count, smallest_cube]
        n = 1

        # Move n up until cube has digits_len digits
        while len(str(n**3)) < digits_len:
            n += 1

        # Process all n where cube has exactly digits_len digits
        while True:
            c = n * n * n
            s = str(c)
            if len(s) > digits_len:
                break

            sig = "".join(sorted(s))
            if sig in seen:
                seen[sig][0] += 1
            else:
                seen[sig] = [1, c]  # first cube is smallest by increasing n

            n += 1

        # Check if any signature reached exactly 5 in this digit-length block
        best = None
        for count, smallest in seen.values():
            if count == 5:
                if best is None or smallest < best:
                    best = smallest

        if best is not None:
            return best

        digits_len += 1

if __name__ == "__main__":
    print(euler_62())  # 127035954683

File: Java/Problem_062.java
import java.util.*;

public class Euler62 {
    static class Info {
        int count;
        long smallest;
        Info(int c, long s) { count = c; smallest = s; }
    }

    static String signature(long x) {
        char[] ch = Long.toString(x).toCharArray();
        Arrays.sort(ch);
        return new String(ch);
    }

    public static void main(String[] args) {
        for (int digitsLen = 1; ; digitsLen++) {
            HashMap<String, Info> seen = new HashMap<>(200000);

            long n = 1;
            while (Long.toString(n*n*n).length() < digitsLen) n++;

            while (true) {
                long c = n*n*n;
                String s = Long.toString(c);
                if (s.length() > digitsLen) break;

                String sig = signature(c);
                Info info = seen.get(sig);
                if (info == null) {
                    seen.put(sig, new Info(1, c)); // first is smallest
                } else {
                    info.count++;
                }
                n++;
            }

            long best = -1;
            for (Info info : seen.values()) {
                if (info.count == 5) {
                    if (best == -1 || info.smallest < best) best = info.smallest;
                }
            }

            if (best != -1) {
                System.out.println(best); // 127035954683
                return;
            }
        }
    }
}
